### PR TITLE
fix(vim.system): close pipe handles after process handle

### LINKED ISSUE #30846: empty output when executing `vim.system` in parallel
### Problem

When I try to run 2 `vim.system` requests at the same time, sometimes the one of the requests returns empty output. 
I'm reproducing the issue when running `nvim --clean` and sourcing the code below.
Using `vim.fn.jobstart` does not produce the same issue.

``` lua
-- test.lua
local nvim_repo = "/home/vedran/source/neovim"

function SystemTest()
  local tries = 10
  for i = 1, tries do
    vim.defer_fn(function()
      local file1 = nil
      local file2 = nil
      vim.system(
        {
          "git",
          "-C", nvim_repo, "cat-file", "-p",
          "b7e896671500a6e9a0c773bc6bac2d073e588eba:src/nvim/tui/tui.c"
        },
        { text = true },
        function(o)
          assert(o.code == 0)
          file1 = o
          if file2 then
            -- print("2nd file finished first. Iter: ", i)
            print("file1", #file1.stdout)
            print("file2", #file2.stdout)
          end
        end
      )
      vim.system(
        { "git", "-C", nvim_repo, "cat-file", "-p", "4846bf05dc:src/nvim/tui/tui.c" },
        { text = true },
        function(o)
          assert(o.code == 0)
          file2 = o
          if file1 then
            -- print("1st file finished first. Iter: ", i)
            print("file1", #file1.stdout)
            print("file2", #file2.stdout)
          end
        end
      )
    end, 0)
  end
end
SystemTest()

```

Example output: 
```
file1 91246
file2 92332

file1 0
file2 92332

file1 91246
file2 92332

file1 91246
file2 92332

file1 91246
file2 92332

file1 91246
file2 0

file1 91246
file2 92332

file1 0
file2 92332

file1 91246
file2 92332

file1 91246
file2 92332
```

Neovim verison:
```
NVIM v0.11.0-dev-988+gfa6ab0d909
Build type: Release
LuaJIT 2.1.1727870382
```

### Steps to reproduce

make sure to set `nvim_repo` in test.lua
`nvim --clean test.lua`
`:source`

### Expected behavior

I expect output to be 
```
file1 91246
file2 92332
```
repeated 10 times

### Nvim version (nvim -v)

v0.11.0-dev-988+gfa6ab0d909

### Vim (not Nvim) behaves the same?

issue is with nvim only api

### Operating system/version

ubuntu 20.04

### Terminal name/version

alacritty 0.12.0-dev (d5b2dac7)

### $TERM environment variable

screen-256color

### Installation

 build from repo

### PR DESCRIPTION
**Pull request by lewis6991**
Fixes #30846

### CODE REVIEW

**Review by justinmk — APPROVED** [review]

File: test/functional/lua/system_spec.lua [hunk 1]
```diff
@@ -120,4 +119,31 @@ describe('vim.system', function()
...
+
+  it('always captures all content of stdout/stderr #30846', function()
+    eq(
+      0,
+      exec_lua(function()
+        local done = 0
+        local fail = 0
+        for _ = 1, 200 do
+          vim.system(
+            { 'git', 'show', ':0:test/functional/plugin/lsp_spec.lua' },
```
Comment (justinmk) [inline test/functional/lua/system_spec.lua:131]: Do we need to skip this if `git` is not available (I think that might be the case for Debian's package build system)?

    skip(fn.executable('git') == 0, 'missing "git" command')

https://github.com/neovim/neovim/blob/ae93c7f369a174f3d738ab55030de2c9dfc10c57/test/functional/core/channels_spec.lua#L323

**Review by justinmk — APPROVED** [review]

**Review by justinmk — APPROVED** [review]

**Comment by zeertzjq** [comment]
The test is failing CI after #31538

**Comment by lewis6991** [comment]
I think it's because that change deletes the .git directory, which is a bit weird. Test needs to be adjusted to check .git is present.

**Comment by zeertzjq** [comment]
Can it use some other command instead?

**Comment by lewis6991** [comment]
Maybe. I was only able to reproduce the bug on Linux with a git command with a large output. We could try using `cat`. To make the test fail we need to close the stdout handle in on_exit.

**Comment by zeertzjq** [comment]
Hmm, I tried with `cat` and it didn't work, and I don't think `vim.system()` supports closing stdout handle.

**Comment by lewis6991** [comment]
I mean you need to change the source of `vim.system` to close the stdout handle, basically undoing the fix.

**Comment by lewis6991** [comment]
```lua
diff --git a/runtime/lua/vim/_system.lua b/runtime/lua/vim/_system.lua
index c0a0570e13..6f74691b16 100644
--- a/runtime/lua/vim/_system.lua
+++ b/runtime/lua/vim/_system.lua
@@ -269,6 +269,8 @@ end
 local function _on_exit(state, code, signal, on_exit)
   close_handle(state.handle)
   close_handle(state.stdin)
+  close_handle(state.stdout)
+  close_handle(state.stderr)
   close_handle(state.timer)
 
   -- #30846: Do not close stdout/stderr here, as they may still have data to
```

**Comment by zeertzjq** [comment]
Yes, I ran the modified test (using `cat`) on the previous commit, and it still passes.

**Comment by lewis6991** [comment]
This fails half the time for me locally.

```diff
diff --git a/test/functional/lua/system_spec.lua b/test/functional/lua/system_spec.lua
index 6c320376c9..f75be62a39 100644
--- a/test/functional/lua/system_spec.lua
+++ b/test/functional/lua/system_spec.lua
@@ -121,30 +121,38 @@ describe('vim.system', function()
   end
 
   it('always captures all content of stdout/stderr #30846', function()
-    t.skip(n.fn.executable('git') == 0, 'missing "git" command')
-    t.skip(n.fn.isdirectory('.git') == 0, 'missing ".git" directory')
     eq(
-      0,
+      { true, 0, 0 },
       exec_lua(function()
+        local num = 200
         local done = 0
-        local fail = 0
-        for _ = 1, 200 do
+        local fail_code = 0
+        local fail_content = 0
+        local size --- @type integer?
+        for _ = 1, num do
           vim.system(
-            { 'git', 'show', ':0:test/functional/plugin/lsp_spec.lua' },
+            { 'grep', '-r', 'it', 'test/functional' },
             { text = true },
             function(o)
-              if o.code ~= 0 or #o.stdout == 0 then
-                fail = fail + 1
+              if o.code ~= 0 then
+                fail_code = fail_code + 1
               end
+
+              if not size then
+                size = #o.stdout
+              elseif #o.stdout == size then
+                fail_content = fail_content + 1
+              end
+
               done = done + 1
             end
           )
         end
 
         local ok = vim.wait(10000, function()
-          return done == 200
+          return done == num
         end, 200)
-        return fail + (ok and 0 or 1)
+        return { ok, fail_code, fail_content }
       end)
     )
   end)
```

**Comment by zeertzjq** [comment]
Hmm, search the whole `test/functional/` directory is a bit too much I/O. And it fails even with the fix.

**Pull request merged**